dmriprep/utils/vectors.py

```diff
@@ -0,0 +1,295 @@
+
+    Returns
+    -------
+    gtab_file : str
+        File path to pickled DiPy gradient table object.
+    gtab : Obj
+        DiPy object storing diffusion gradient information.
+    final_fbval : str
+        File path of saved final b-values if `final` is True.
+    final_fbvec : str
+        File path of saved final b-vectors if `final` is True.
+    """
+    import os
+    from dipy.io import save_pickle, read_bvals_bvecs
+    from dipy.core.gradients import gradient_table
+
+    # Read b-values and b-vectors from disk
+    # Files can be either ‘.bvals’/’.bvecs’ or ‘.txt’ or ‘.npy’ (containing
+    # arrays stored with the appropriate values).
+    if fbval and fbvec:
+        bvals, bvecs = read_bvals_bvecs(fbval, fbvec)
+    else:
+        raise OSError("Either bvals or bvecs files not found (or rescaling failed)!")
+
+    # Creating the gradient table
+    gtab = gradient_table(bvals, bvecs, b0_threshold=b0_thr)
+
+    # Show info
+    print(gtab.info)
+
+    # Save bval/bvec text file equivalents in the case that these are the final
+    # versions of our gradient vectors
+    if final is True:
+        final_bval_path = "%s%s" % (vector_dir, '/final_bval.bval')
+        final_bvec_path = "%s%s" % (vector_dir, '/final_bvec.bvec')
+        np.savetxt(final_bval_path, bvals.astype('int'), fmt='%d', newline=' ')
+        np.savetxt(final_bvec_path, bvecs.T.astype('float'), fmt='%8f')
+        # Save gradient table to pickle
+        gtab_file = "%s%s" % (vector_dir, "/gtab_final.pkl")
+        save_pickle(gtab_file, gtab)
+    else:
+        def next_path(path_pattern):
+            i = 1
+            while os.path.exists(path_pattern % i):
+                i = i * 2
+            a, b = (i // 2, i)
+            while a + 1 < b:
+                c = (a + b) // 2 # interval midpoint
+                a, b = (c, b) if os.path.exists(path_pattern % c) else (a, c)
+            return path_pattern % b
+        # Save gradient table to pickle
+        gtab_file = next_path(vector_dir + "/gtab_%s.pkl")
+        save_pickle(gtab_file, gtab)
+        final_bval_path = None
+        final_bvec_path = None
+
+    return gtab_file, gtab, final_bval_path, final_bvec_path
+
+
+def rescale_bvec(bvec):
+    """
+    Normalizes b-vectors to be of unit length for the non-zero b-values. If the
+    b-vector row reflect a B0, the vector is untouched.
+
+    Parameters
+    ----------
+    bvec : m x n 2d array
+        Raw b-vectors array.
+
+    Returns
+    -------
+    bvec_rescaled : m x n 2d array
+        Unit-normed b-vectors array.
+    """
+    # Enforce proper dimensions
+    bvec = bvec.T if bvec.shape[0] == 3 else bvec
+
+    # Normalize values not close to norm 1
+    bvec = [b / np.linalg.norm(b) if not np.isclose(np.linalg.norm(b), 0) else b for b in bvec]
+    norms = np.linalg.norm(bvec, axis=-1)
+    norms[np.isclose(norms, 0)] = 1
+    bvec_rescaled = bvec / norms[:, None]
+    return bvec_rescaled
+
+
+def rescale_bval(bval, bvec):
```

Comment:
I think you want to merge both rescalings.

Reply:
Hmm, the only reason that I am hesitant on this one is that a user may not wish to rescale the bval. For instance, mrtrix allows the user to disable it altogether: See: https://mrtrix.readthedocs.io/en/latest/concepts/dw_scheme.html#b-value%20scaling

Reply:
Please chime in here - https://github.com/bids-standard/bids-specification/pull/352#discussion_r337617411

But for dMRIPrep I would go with absolute b-vals without any doubt.

Reply:
Whatever our dmriprep developers agree is best, I'm happy to go with. Bear in mind that this is a separate correction from b-value rescaling performed based on b order of magnitude.

I am a bit concerned that like mrtrix outlines, mandatory b-val rescaling by vector amplitude, while beneficial in the vast majority of cases, could introduce "minor rounding errors in the components of the direction vectors" in some cases. 

If we do decide to make b-value vector rescaling optional, I could just add a `bval_scaling` flag, and make `True` the default?

Reply:
Also, I should note that any scaling of the b-vals gets rounded to the nearest integer (99/100 these are identical to the original). I'm still happy to disable b-val scaling altogether though if by 'absolute b-values' you meant the original, untouched?